compiler/src/org.graalvm.compiler.replacements/src/org/graalvm/compiler/replacements/SnippetTemplate.java

```diff
@@ -374,9 +374,9 @@ protected Lazy lazy() {
 
         protected int nextParamIdx;
 
-        public Arguments(SnippetInfo info, GuardsStage guardsStage, LoweringTool.LoweringStage loweringStage) {
+        public Arguments(SnippetInfo info, GuardsStage guardsStage, LoweringTool.LoweringStage loweringStage, boolean expandsVolatileMembarsLate) {
```

Comment:
I think the `loweringStage` argument is always coming from a call to `org.graalvm.compiler.nodes.spi.LoweringTool.getLoweringStage()` in which case the `LoweringTool` object can be passed to this constructor and the constructor can extract the `loweringStage` and `expandsVolatileMembarsLate` values itself.

Reply:
I made that change and pushed an updated patch.

Reply:
Looks good.

Reply:
Thanks for your help with this.

Reply:
No problem. Note that @sanzinger will have to make some changes in the enterprise code base to adapt to these changes so probably won't get around to re-submitting for merge until next week (he's currently on vacation).

Reply:
> I think the loweringStage argument is always coming from a call to org.graalvm.compiler.nodes.spi.LoweringTool.getLoweringStage()

This assumption is wrong. The enterprise implementation uses snippets outside of lowerings as well, where we don't have any `LoweringTool`.
I think the fact that "volatile field reads/writes with barriers" is a property of the entire graph and should be attached there. I could either think of a member in `StructuredGraph` or even an additional property such as `StoreFieldNode.requireMembar` or `LoadFieldNode.requireMembar`.